tools/data_integrity/score_instrument.php

```diff
@@ -0,0 +1,241 @@
+ *
+ * The 'all' option scores all existing records of an instrument if and only if
+ *  - They belong to an active timepoint
+ *  - The data-entry status of the timepoint is set to `complete`
+ *  - The administration of the timepoint is NOT `none`
+ *
+ * Note: This tool can also be used to debug the scoring algorithms for development.
+ *
+ * Limitation: This tool does not reset or nullify the score of an instrument
+ * which was previously scored but no longer meets the criteria above (i.e. if the
+ * administration was changed from `all` to `none`, the score will not be removed).
+ *
+ * @package behavioural
+ */
+
+require_once __DIR__."/../generic_includes.php";
+
+// LOGGING
+$dir = __DIR__ . "/../logs/";
+if (!is_dir($dir)) {
+    mkdir($dir);
+}
+$today   = getdate();
+$date    = strftime("%Y-%m-%d_%H:%M");
+$logPath = "$dir/score_instrument.$date.log";
+$logfp   = fopen($logPath, 'a');
+
+if (!$logfp) {
+    printError(
+        "No logs can be generated, path:$logPath ".
+        "does not exist or can not be written to.\n"
+    );
+}
+
+/**
+ * HELP SCREEN
+ * display and stop processing if action=help
+ */
+if (empty($argv[1])
+    || $argv[1] == 'help'
+    || in_array('-h', $argv, true)
+    || ($argv[1] === 'all' && isset($argv[2]))
+    || ($argv[1] !== 'all' && empty($argv[2]))
+    || (isset($argv[2]) && !in_array($argv[2], array('all', 'one'), true))
+    || (isset($argv[2]) && $argv[2]=='one'
+    && (empty($argv[3]) || empty($argv[4])))
+) {
+    printOut(
+        "Usage:
+        score_instrument.php [help | -h]
+            -> displays this message
+        score_instrument.php <test_name> one <candID> <sessionID>
+            -> scores test only for this candidate
+        score_instrument.php <test_name> all
+            -> scores test for all candidates meeting the necessary criteria
+        score_instrument.php all
+            -> scores all instruments for all candidates meeting the necessary criteria \n\n"
+    );
+    return;
+}
+
+/**
+ * get cmd-line arguments
+ */
+// get $action argument
+$test_name = $argv[1];
+// get $action argument
+$action ='';
+if (isset($argv[2])) {
+    $action = strtolower($argv[2]);
+}
+// CandID
+if (!empty($argv[3])) { $candID = $argv[3];
+}
+// sessionID
+if (!empty($argv[4])) { $sessionID = $argv[4];
+}
+
+// include instrument class
+if(strtolower($test_name) != 'all') {
+    if (!is_file(__DIR__."/../../project/instruments/NDB_BVL_Instrument_$test_name.class.inc")
+        && !is_file(__DIR__."/../../project/instruments/$test_name.linst")
+    ) {
+        printError("Included file does not exist (".__DIR__."/../../project/instruments/NDB_BVL_Instrument_$test_name.class.inc)");
+        return false;
+    }
+}
+
+// check that the $test_name is a valid instrument
+if($test_name== 'all') {
+    $query = "SELECT DISTINCT test_name FROM test_battery WHERE Active='Y'";
+} else {
+    $query = "SELECT DISTINCT test_name FROM test_battery WHERE Active='Y' AND Test_name = :tnm";
```

Comment:
test_battery doesn't tell you whether it's a valid instrument, it tells you if it's part of a battery. Shouldn't this be checking test_names instead?

Reply:
good point I'll try to figure out what this check is trying to do and either correct the documentation or change the queries

Reply:
@driusan I investigated...

the code doesn't use anything from `test_battery` other than the `test_name` field however my suspicion is that the `test_battery` table is being used because of the filter `Active=Y` as in "Only fix scoring for instruments that are part of an active test battery"

I am really not sure how much of a change in behaviour it is to point it at the "test_names" table instead... my guess is because `test_names` does not have an active/inactive flag, people have been inactivation "tests" from the `test_battery` table instead !!

thoughts?

Reply:
Can you add a comment explaining the logic of only updating scores for Active test batteries? That makes sense to me.

Reply:
no I cant :s 

Loris meeting

Reply:
DONE !!